pkg/asset/machines/aws/worker.go

```diff
@@ -30,6 +37,83 @@ type MachineConfig struct {
 	Machine     types.AWSMachinePoolPlatform
 }
 
+func createWorkerMachineSets(config *WorkerConfig) metav1.List {
+	list := metav1.List{}
+	list.Kind = "List"
+	items := make([]runtime.RawExtension, 0)
+	for i, instance := range config.Instances {
+		ms := clusterapi.MachineSet{}
+		ms.Kind = "MachineSet"
+		ms.Name = fmt.Sprintf("%s-worker-%d", config.ClusterName, i)
+		ms.Namespace = "openshift-cluster-api"
+		ms.Labels = map[string]string{
+			"sigs.k8s.io/cluster-api-cluster":      config.ClusterName,
+			"sigs.k8s.io/cluster-api-machine-role": "worker",
+			"sigs.k8s.io/cluster-api-machine-type": "worker",
+		}
+		ms.Spec = clusterapi.MachineSetSpec{}
+		ms.Spec.Replicas = new(int32)
+		*ms.Spec.Replicas = int32(instance.Replicas)
+		ms.Spec.Selector = metav1.LabelSelector{}
+		ms.Spec.Selector.MatchLabels = map[string]string{
+			"sigs.k8s.io/cluster-api-machineset": "worker",
+			"sigs.k8s.io/cluster-api-cluster":    config.ClusterName,
+		}
+
+		template := clusterapi.MachineTemplateSpec{}
+		template.Labels = map[string]string{
+			"sigs.k8s.io/cluster-api-machineset":   "worker",
+			"sigs.k8s.io/cluster-api-cluster":      config.ClusterName,
+			"sigs.k8s.io/cluster-api-machine-role": "worker",
+			"sigs.k8s.io/cluster-api-machine-type": "worker",
+		}
+		template.Spec = clusterapi.MachineSpec{}
+		provider := awsprovider.AWSMachineProviderConfig{}
+		provider.APIVersion = "aws.cluster.k8s.io/v1alpha1"
+		provider.InstanceType = config.Machine.InstanceType
+		provider.AMI = awsprovider.AWSResourceReference{ID: new(string)}
+		*provider.AMI.ID = config.AMIID
+		provider.Subnet = awsprovider.AWSResourceReference{
+			Filters: []awsprovider.Filter{
+				{
+					Name:   "tag:Name",
+					Values: []string{fmt.Sprintf("%s-worker-%s", config.ClusterName, instance.AvailabilityZone)},
+				},
+			},
+		}
+		provider.Placement = awsprovider.Placement{Region: config.Region, AvailabilityZone: instance.AvailabilityZone}
+		provider.IAMInstanceProfile = &awsprovider.AWSResourceReference{ID: new(string)}
+		*provider.IAMInstanceProfile.ID = fmt.Sprintf("%s-worker-profile", config.ClusterName)
+		provider.Tags = make([]awsprovider.TagSpecification, 0)
+		for tagName, tagValue := range config.Tags {
+			provider.Tags = append(provider.Tags, awsprovider.TagSpecification{Name: tagName, Value: tagValue})
+		}
+		provider.SecurityGroups = []awsprovider.AWSResourceReference{
+			{
+				Filters: []awsprovider.Filter{
+					{
+						Name:   "tag:Name",
+						Values: []string{fmt.Sprintf("%s_worker_sg", config.ClusterName)},
+					},
+				},
+			},
+		}
+		provider.UserDataSecret = &corev1.LocalObjectReference{Name: "worker-user-data"}
+
+		template.Spec.ProviderConfig = clusterapi.ProviderConfig{Value: &runtime.RawExtension{Object: &provider}}
+		template.Spec.Versions = clusterapi.MachineVersionInfo{
+			Kubelet:      "",
+			ControlPlane: "",
```

Comment:
That is how it was in the template. I am guessing we do want to write the correct values at some point. Next to no harm for now.

Reply:
> That is how it was in the template.

Presumably because the upstream type is missing some `omitempty`.

> I am guessing we do want to write the correct values at some point. Next to no harm for now.

I'd rather address that with `FIXME` comments in the code and/or GitHub issues to track the planned future work.  Leaving empty strings here doesn't seem like an obvious way to track future work.